STM32: rework timer frequency protection

tpruvot opened the pull request:
STEP_TIMER_NUM, TEMP_TIMER_NUM and PULSE_TIMER_NUM are not the indexes returned by the STM32 framework get_timer_index() function. Same thing now with the recently renamed MF_TIMER_xxx constants

In STM32 HAL these constants are defined by Marlin for its own container of custom timers, timer_instance[] which has only 2 records.

follow up for #23125

descipher commented:
Please consider:

https://github.com/MarlinFirmware/Marlin/blob/a7549aac5c5efc7097d6ebe69b2d46b3e7acbfab/Marlin/src/module/stepper.cpp#L2772

and 

https://github.com/MarlinFirmware/Marlin/blob/a7549aac5c5efc7097d6ebe69b2d46b3e7acbfab/Marlin/src/module/temperature.cpp#L2378

These calls use the STEP_TIMER_NUM and TEMP_TIMER_NUM refs.

descipher commented:
So it is only true for the maple lib:
''
'#ifndef STEP_TIMER_NUM
  #if defined(MCU_STM32F103CB) || defined(MCU_STM32F103C8)
    #define STEP_TIMER_NUM      4  // For C8/CB boards, use timer 4
  #else
    #define STEP_TIMER_NUM      5  // for other boards, five is fine.
  #endif
#endif
#ifndef PULSE_TIMER_NUM
  #define PULSE_TIMER_NUM       STEP_TIMER_NUM
#endif
#ifndef TEMP_TIMER_NUM
  #define TEMP_TIMER_NUM        2  // Timer Index for Temperature
  //#define TEMP_TIMER_NUM      4  // 2->4, Timer 2 for Stepper Current PWM
#endif
'''

descipher commented:
Why would we not keep this consistent? Array Index in one vs actual timer in the other?

descipher commented:
The same issue exists with the __STM32F1__ HAL the actual timers are not TEMP_TIMER_NUM or STEP_TIMER_NUM.

descipher commented:
Thanks for the followup.

tpruvot commented:
> Please consider:https://github.com/MarlinFirmware/Marlin/blob/a7549aac5c5efc7097d6ebe69b2d46b3e7acbfab/Marlin/src/module/stepper.cpp#L2772
> 
> and
> https://github.com/MarlinFirmware/Marlin/blob/a7549aac5c5efc7097d6ebe69b2d46b3e7acbfab/Marlin/src/module/temperature.cpp#L2378
> 
> These calls use the STEP_TIMER_NUM and TEMP_TIMER_NUM refs.

yes... but HAL_timer_start() use the marlin array of HardwareTimer

and yes STEP_TIMER_NUM  5 is only for the F1 HAL, i agree the STM32 code can be confusing (and confused me before) with the same defines constants used diffently... this PR is only for STM32 HAL, not maple/F1 HAL

descipher commented:
Tested. HAL_STM32 passed 👍 
Maple HAL failed, due to it using different pin mapping to timers, also it sets the wrong frequency .. lol (500Hz)
I am looking at the Maple issue. It's related to the library and of course mappings are a mess.

tpruvot commented:
Thanks for the report. Indeed my current printer can't use pwm on the maple framework due to the missing alternate function pin mapping, but F1 timer HAL code seems more "simple" to use, at least for the timer numbers...

descipher commented:
I think AF mapping is not really missing. It just cannot do it based on the method used in the timer_dev structs and other device structs. There is no pointer assignment for the alternate pin in the map table.
e.g.
```
{&gpiod,    NULL, NULL,  11, 0, ADCx} , /* PD11  FSMC_A16 */
{&gpiod,    NULL, NULL,  12, 0, ADCx} , /* PD12  FSMC_A17 */
{&gpiod,    NULL, NULL,  13, 0, ADCx} , /* PD13  FSMC_A18 */	
{&gpiod,    NULL, NULL,  14, 0, ADCx} , /* PD14  FSMC_D0 */
{&gpiod,    NULL, NULL,  15, 0, ADCx} , /* PD15  FSMC_D1 */
```
Could be set as follows:

```
{&gpiod,    NULL, NULL,  11, 0, ADCx} , /* PD11  FSMC_A16 */
{&gpiod, &timer4, NULL,  12, 1, ADCx} , /* PD12  FSMC_A17 */
{&gpiod, &timer4, NULL,  13, 2, ADCx} , /* PD13  FSMC_A18 */	
{&gpiod, &timer4, NULL,  14, 3, ADCx} , /* PD14  FSMC_D0 */
{&gpiod, &timer4, NULL,  15, 4, ADCx} , /* PD15  FSMC_D1 */
```

tpruvot commented:
i didnt try, but maybe if i remove the standard timer4 pins in the variant, i can then set it on PD12 & 13. May need a afio_remap call then...

descipher commented:
I would not recommend Maple after looking deeper at it.

thinkyhead commented:
The value returned by `get_timer_index` is not necessarily going to be the timer number minus one, since `timer_index_t` is an `enum` containing only enabled timers, with skips and gaps. So I changed the `index` test to compare the returned `get_timer_index` value against the `enum` for the timer index, rather than the timer index minus one.

descipher commented:
You should be good using the index value based on the libs methods.
```
/**
  * @brief  HardwareTimer destructor
  * @retval None
  */
HardwareTimer::~HardwareTimer()
{
  uint32_t index = get_timer_index(_timerObj.handle.Instance);
  disableTimerClock(&(_timerObj.handle));
  HardwareTimer_Handle[index] = NULL;
  _timerObj.__this = NULL;
}

/**
  * @brief  return timer index from timer handle
  * @param  htim : one of the defined timer
  * @retval None
  */
timer_index_t get_timer_index(TIM_TypeDef *instance)
{
  timer_index_t index = UNKNOWN_TIMER;

#if defined(TIM1_BASE)
  if (instance == TIM1) {
    index = TIMER1_INDEX;
  }
#endif
```

tpruvot commented on Marlin/src/HAL/STM32/fast_pwm.cpp:
```diff
@@ -25,69 +25,66 @@
+
+  // Protect used timers.
+  #define __TINDEX(T) TIMER##T##_INDEX // Timer indexes from TIMER#_INDEX enum which depends on TIM#_BASE defines.
+  #define _TINDEX(T) __TINDEX(T)
```
could be useful to have this macro in HAL timers.h, and maybe with a more explicit name like TIMER_INDEX(NUM)

tpruvot commented on Marlin/src/HAL/STM32/fast_pwm.cpp:
```diff
@@ -25,63 +25,59 @@
+  #endif
+  #if defined(PULSE_TIMER) && MF_TIMER_PULSE != MF_TIMER_STEP
+    if (index == TIMER_INDEX(PULSE_TIMER)) return;
+  #endif
```
im not sure this last one is useful here, or correctly coded in STM32 HAL (see timers.h and timers.cpp)

```
#define NUM_HARDWARE_TIMERS 2
```
prevents the use/define of it...

```
#ifndef MF_TIMER_PULSE
  #define MF_TIMER_PULSE        MF_TIMER_STEP
#endif
```

also for the other ones... 
```
#ifndef MF_TIMER_STEP
  #define MF_TIMER_STEP         0  // Timer Index for Stepper
#endif
```

for STM32... there is no meaning of using these ifndef ... because its just offsets in a marlin array, unrelated to timer selection so.

thinkyhead replied:
Yes, these timers should always exist, and no overrides have ever been needed or proposed, so the HALs that have these values fixed and not related to platform timer IDs should drop the `#ifdef` wrappers, which give the incorrect impression that they can be overridden in some way.

tpruvot commented on Marlin/src/HAL/STM32/fast_pwm.cpp:
```diff
@@ -25,63 +25,59 @@
-    if (timer_freq[index] == 0 ) {                    // If the timer is unconfigured and no freq is set then default PWM_FREQUENCY.
+    if (timer_freq[index] == 0) {                     // If the timer is unconfigured and no freq is set then default PWM_FREQUENCY.
       timer_freq[index] = PWM_FREQUENCY;
       set_pwm_frequency(pin_name, timer_freq[index]); // Set the frequency and save the value to the assigned index no.
```
here: set_pwm_frequency(pin_name, PWM_FREQUENCY); is enough... 
timer_freq[index] = f_desired; is made in set_pwm_frequency()

tpruvot commented on Marlin/src/HAL/STM32/timers.h:
```diff
@@ -41,16 +41,9 @@
-#endif
+#define MF_TIMER_STEP         0  // Timer Index for Stepper
+#define MF_TIMER_TEMP         1  // Timer Index for Temperature
+#define MF_TIMER_PULSE        MF_TIMER_STEP
```
maybe some comment should be added to explain these values are not tunable....

tpruvot replied:
unlike some other HALs....

ellensp commented:
The simulator is also broken with error: ‘STEP_TIMER_NUM’ was not declared in this scope

tpruvot commented:
hmm no, there is no more STEP_TIMER_NUM ref in git...

descipher commented:
> hmm no, there is no more STEP_TIMER_NUM ref in git... (after this PR at least)

Yes, I see Scott has refactored them for all HALs so a merge would be in order for any in progress PRs.

tpruvot committed:
STM32: rework timer frequency protection

STEP_TIMER_NUM, TEMP_TIMER_NUM and PULSE_TIMER_NUM are not the indexes returned by the STM32 framework get_timer_index()

In STM32 HAL these constants are defined by Marlin for its own container of custom timers, timer_instance[] which has only 2 or 3 records.

Also initialise the boolean variable and remove some extra spaces...

follow up for #23125

tpruvot commented:
rebased & squashed as a single commit (easier to cherry pick)

descipher commented:
> rebased & squashed as a single commit (easier to cherry pick)

Indeed, looks good.

tpruvot committed:
Include pre-post-merge mess
